### ISSUE TITLE
SSH auth

### ISSUE THREAD

**Issue by miekg** [miekg/gitopper#16]
Was thinking about doing auth, and eyeing TLS, Kerberos, etc. But then I thought, wait, how do I login on my home servers? And that SSH. So I think it makes sense to default to SSH and a Go pkg exists for it.

It may be possible to shoehorn the REST stuff into SSH or rewrite that bit as well, so the whole proto becomes SSH. Added benefit IMHO is no TLS certs (and validity of those), IP only hosts and piggybacking on a well understood proto for sysadmins.

I.e. you auth with ssh, then we check if you're in the "special" group and if so you can manipulate gitopper.

Catch-22: how to setup the ssh infra to get all this  going? As said in design doc - think a custom image should solve this by having some sane default that make sense for you org

**Comment by miekg** [comment]
https://pkg.go.dev/golang.org/x/crypto/ssh#example-Client.Listen 
but unsure how you would hook into there was our own code....

**Comment by miekg** [comment]
does it make sense to include all auth by ourselves, i.e. public key auth and carry the keys in the config (that we pull down anyway) - that means we don't depend on system files at all

**Comment by miekg** [comment]
This Go program allows me to ssh into it. I can intercept various things and then call the json creation methods that are now listed in routes.go
~~~ go
package main

import (
	"fmt"
	"io"
	"io/ioutil"
	"log"

	"github.com/gliderlabs/ssh"
)

func main() {
	// generate persistent key
	ssh.Handle(func(s ssh.Session) {
		io.WriteString(s, fmt.Sprintf("Hello %s %v\n", s.User(), s.Command()))
	})

	log.Println("starting ssh server on port 2222...")
	ssh.ListenAndServe(":2222", nil,
		ssh.PublicKeyAuth(func(ctx ssh.Context, key ssh.PublicKey) bool {
			data, _ := ioutil.ReadFile("/home/miek/.ssh/id_ed25519.pub")
			allowed, _, _, _, _ := ssh.ParseAuthorizedKey(data)
			return ssh.KeysEqual(key, allowed)
		}),
	)
}
~~~

**Comment by miekg** [comment]
We need to store the host key somewhere however - or make the client not care when it connects. I'm inclined to go for the latter, because once you store state it becomes convenient to store more of it.

Also if we pursue this we need to extend this config with pub keys of admins(?)

~~~ toml
[[keys]]
user = "miek"
path = "keys/id_ed25519.pub"
~~~
where those keys also live in the config repo of gitopper.

**Comment by miekg** [comment]
closed via #17
